Packs/PAN-OS/Integrations/Panorama/Panorama.py

```diff
@@ -7011,6 +7154,1999 @@ def panorama_install_file_content_update_command(args: dict):
+                yield value
+
+    def active_devices(self, filter_str: str = None) -> Generator[Union[Firewall, Panorama], None, None]:
+        """
+        Yields active devices in the topology
+        """
+        # If the ha_active_devices dict is not empty, we have gotten HA info from panorama.
+        # This means we don't need to refresh the state.
+        for device in self.all(filter_str):
+            if self.ha_active_devices:
+                if device.serial in self.ha_active_devices:
+                    yield device
+            else:
+                status = device.refresh_ha_active()
+                if status == "active" or not status:
+                    yield device
+
+    def active_top_level_devices(self, device_filter_string: str = None):
+        """
+        Yields active top level devices
+        """
+        active_top_level_devices = [x for x in self.top_level_devices() if
+                                    x in self.active_devices(device_filter_string)]
+        return active_top_level_devices
+
+    @staticmethod
+    def filter_devices(devices: dict, filter_str: Optional[str] = None):
+        """Filters a list of devices to find matching entries based on the string"""
+        # Exact match based on device serial number
+        if not filter_str:
+            return devices
+
+        if filter_str in devices:
+            return {
+                filter_str: devices.get(filter_str)
+            }
+
+        device: PanDevice
+        for serial, device in devices.items():
+            if device.hostname == filter_str:
+                return {
+                    serial: device
+                }
+
+    def firewalls(self, filter_string: str = None) -> Generator[Firewall, None, None]:
+        """Returns an iterable for each firewall in the topology"""
+        firewall_objects = Topology.filter_devices(self.firewall_objects, filter_string)
+        if not firewall_objects:
+            raise DemistoException("Filter string returned no devices known to this topology.")
+
+        for firewall in firewall_objects.values():
+            yield firewall
+
+    def all(self, filter_string: str = None) -> Generator[Union[Firewall, Panorama], None, None]:
+        all_devices = {**self.firewall_objects, **self.panorama_objects}
+        all_devices = Topology.filter_devices(all_devices, filter_string)
+        # Raise if we get an empty dict back
+        if not all_devices:
+            raise DemistoException("Filter string returned no devices known to this topology.")
+
+        for device in all_devices.values():
+            yield device
+
+    def get_by_filter_str(self, filter_string: str = None) -> dict:
+        all_devices = {**self.firewall_objects, **self.panorama_objects}
+        all_devices = Topology.filter_devices(all_devices, filter_string)
+        return all_devices
+
+    @classmethod
+    def build_from_string(cls, hostnames: str, username: str, password: str, api_key: str = None):
```

Comment:
why don't you just build it from the constructor of ```Topology``` class instead of using a ```classmethod``` decorator?

Reply:
when I wrote it I was thinking I might have other ways of building the topology, so it makes sense to split them into class methods for example `build_from_string`, `build_from_device`, `build_from_integration_context`.

Reply:
Very nice thinking, I like your idea, keep those class methods. 
In addition for every method of that kind, it would be nice if you can add a doc string with an example explaining that.
For example what does build from string means? and what does build from device mean? I would be happy to understand the context better.

Overall incredible work, well done!

Reply:
Updated this docstring slightly to provide this context